onnxruntime/contrib_ops/rocm/bert/tunable_op.h

```diff
@@ -3,70 +3,116 @@
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
 #include <vector>
-#include <hip/hip_runtime.h>
+
+#include "core/common/common.h"
 #include "contrib_ops/rocm/bert/util.h"
 
 namespace onnxruntime {
 namespace contrib {
 namespace rocm {
 
 struct OpParams {
+  OpParams() : stream{} {}
   explicit OpParams(hipStream_t stream) : stream(stream) {}
-  virtual std::string signature() const = 0;
   hipStream_t stream;
 };
 
+// A type erased Callable wrapper. We could have used std::function<Status<const ParamT*>> here. However, std::function
+// requires the callable object to be CopyConstructible and CopyAssignable. This is not suitable for move only functor
+// or move captured lambda. So we create a simple wrapper for our purpose here.
+//
+// Then an Op is Status(const ParamT*), that is, a callable accepts a const ParamT* and returns a Status.
+// This means that it can be either a free function, a functor or a lambda.
+template <typename ParamT>
 class Op {
  public:
-  Op() : repeats_(100) {}
+  template <typename T>
+  explicit Op(T&& c) : callable_{std::make_unique<CallableImpl<T>>(std::forward<T>(c))} {}
+  Status operator()(const ParamT* param) { return (*callable_)(param); }
 
-  virtual void Run(const OpParams*) = 0;
+ private:
+  struct ICallbale {
+    virtual ~ICallbale() = default;
+    virtual Status operator()(const ParamT*) = 0;
+  };
 
-  void SetRepeats(int n) {
-    repeats_ = n;
-  }
+  template <typename T>
+  struct CallableImpl : ICallbale {
+    explicit CallableImpl(T&& c) : c_{std::move(c)} {}
+    Status operator()(const ParamT* param) override { return c_(param); }
 
-  float Profile(const OpParams* op_params) {
-    // warm up
-    for (int i = 0; i < 5; i++) {
-      Run(op_params);
-    }
-    timer_.Start();
-    for (int i = 0; i < repeats_; i++) {
-      Run(op_params);
-    }
-    timer_.End();
-    return timer_.time()/repeats_;
-  }
-
-  virtual ~Op() {}
+   private:
+    T c_;
+  };
 
- private:
-  Timer timer_;
-  int repeats_;
+  std::unique_ptr<ICallbale> callable_;
 };
 
+// NOTE: onnxruntime's status currently do not have a StatusCode::UNSUPPORTED. Currently, we do not want to extend the
+// enum. So we reuse StatusCode::INVALID_ARGUMENT for this purpose. It can be interpreted as "The input argument is not
+// valid for this specialized kernel implementation.". This semantic is crucial for the tuning mechanism.
+#define TUNABLE_OP_RETURN_UNSUPPOTED_ARGUMENT_IF(condition, ...)   \
+  do {                                                             \
+    if (condition) {                                               \
+      return ORT_MAKE_STATUS(NONE, INVALID_ARGUMENT, __VA_ARGS__); \
+    }                                                              \
+  } while (false)
+
+template <typename ParamsT>
 class TunableOp {
  public:
-  explicit TunableOp(int default_id) : default_id_(default_id), tuning_(false) {}
+  static bool IsSupported(Op<ParamsT>& op, const ParamsT* param) {
```

Comment:
Yeah, but be aware they have minor different behavior. `Condition` only checks `param`, but it is hard to **correlate with the impl op**, since they are generally queried at runtime and might change silently due environment or library upgrade. So for us, it  is impossible to maintain a table to identify, say impl `ops_[1]` does not support `Params a`. And for IsSupported, instead of trying to maintain a table, it actually calls into the impl `ops_[1]` with `Params a` and checks the returned `Status` to do the identification. I am wondering if we are actually need the `Condition` here.